src/extension/coqPilot.ts

```diff
@@ -81,70 +75,126 @@ export class CoqPilot {
         shouldCompleteHole: (hole: ProofStep) => boolean,
-        editor: TextEditor
+        editor: TextEditor,
+        abortSignal: AbortSignal
     ) {
-        await window.withProgress(
-            {
-                location: ProgressLocation.Window,
-                title: `${pluginName}: In progress`,
-            },
-            async () => {
-                try {
-                    await this.performSpecificCompletions(
-                        shouldCompleteHole,
-                        editor
-                    );
-                } catch (e) {
-                    if (e instanceof SettingsValidationError) {
-                        e.showAsMessageToUser();
-                    } else if (e instanceof CoqLspStartupError) {
-                        showMessageToUserWithSettingsHint(
-                            EditorMessages.coqLspStartupFailure(e.path),
-                            "error",
-                            `${pluginId}.coqLspServerPath`
-                        );
-                    } else {
-                        showMessageToUser(
-                            e instanceof Error
-                                ? EditorMessages.errorOccurred(e.message)
-                                : EditorMessages.objectWasThrownAsError(e),
-                            "error"
-                        );
-                        console.error(buildErrorCompleteLog(e));
-                    }
-                }
+        if (!this.globalExtensionState.hasActiveSession) {
+            showMessageToUser(EditorMessages.extensionIsPaused, "warning");
+            return;
+        }
+        try {
+            this.healthStatusIndicator.showSpinner();
+
+            await this.performSpecificCompletions(
+                shouldCompleteHole,
+                editor,
+                abortSignal
+            );
+        } catch (e) {
+            if (e instanceof SettingsValidationError) {
+                e.showAsMessageToUser();
+            } else if (e instanceof CoqLspStartupError) {
+                showMessageToUserWithSettingsHint(
+                    EditorMessages.coqLspStartupFailure(e.path),
+                    "error",
+                    `${pluginId}.coqLspServerPath`
+                );
+            } else if (
+                e instanceof CompletionAbortError ||
+                !this.globalExtensionState.hasActiveSession
+            ) {
+                showMessageToUser(EditorMessages.completionAborted, "info");
+            } else {
+                showMessageToUser(
+                    e instanceof Error
+                        ? EditorMessages.errorOccurred(e.message)
+                        : EditorMessages.objectWasThrownAsError(e),
+                    "error"
+                );
+                console.error(buildErrorCompleteLog(e));
             }
-        );
+        } finally {
+            this.healthStatusIndicator.hideSpinner(
+                this.globalExtensionState.hasActiveSession
```

Comment:
Is this code truly reachable with the sessions being active?

Reply:
Yes, I checked that it works as expected